src/vsr/journal.zig

```diff
@@ -2083,6 +2016,13 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
                 current = waiting.next;
                 waiting.next = null;
                 journal.lock_sectors(@as(*Journal.Write, @fieldParentPtr("range", waiting)));
+                // Header sector locks are always an exact match, so there's no need to re-check the
+                // the waiting writes against all other writes.
+                if (range.ring == .headers) {
+                    assert(waiting.locked);
+                    waiting.next = current;
+                    break;
+                }
```

Comment:
This approach of overwriting `waiting.next` (which I assume is guaranteed to be `null` still?) and then early-exiting from the loop makes me a little nervous. I think I'd rather just reuse the same loop code, even at the cost of the early-exist optimization.

Alternatively, you could refactor `lock_sectors` such that there is a variant that can be called when the write already has a queue behind it, and is guaranteed not to collide with an in-flight write.

Reply:
Mhmmmmm, I *did* feel a bit awkward overwriting waiting.next as well. I had briefly tried branching out `lock_sectors` based on `ring == .headers` but it felt a bit unergonomic as well... Let me take another stab at this!

However, I don't think `waiting.next` would be guaranteed to be null, here? IIUC, if we were to remove this branch and just do the below (the older logic), the first header that is queued would be able to take the lock, and then the rest of the ones waiting would be queued on it again after this loop (which is the state that we arrive at by breaking out of the loop early as well).
```C
            var current = range.next;
            range.next = null;
            while (current) |waiting| {
                assert(waiting.locked == false);
                current = waiting.next;
                waiting.next = null;
                journal.lock_sectors(@as(*Journal.Write, @fieldParentPtr("range", waiting)));
            }

```

The only difference is that each queued header write would have to iterate through all of the write IOPs: https://github.com/tigerbeetle/tigerbeetle/blob/a4cb4e87dc889a84eb7e7726a7738ca75e87f808/src/vsr/journal.zig#L1951-L1962

Reply:
`lock_sectors` asserts that `write.range.next == null`, and we set it to `null` immediately before calling `lock_sectors`:

```
                waiting.next = null;
                journal.lock_sectors(@as(*Journal.Write, @fieldParentPtr("range", waiting)));
```

And `lock_sectors` itself doesn't modify its parameter's `next` (though it may append it to a different write).

Reply:
Ah yes, makes sense! Decided to go for the simpler implementation here instead of implementing a variant of `lock_sectors`: https://github.com/tigerbeetle/tigerbeetle/pull/2273/commits/569a1e3f342c7c6ff8b243d5953df6b5641cd508.